### PR TITLE
Posterior saving updates

### LINKED ISSUE #606: load posterior subclasses
@PierreBoyeau how should we handle loading posterior subclasses like TotalPosterior and JVAEPosterior?

### LINKED ISSUE #612: load posterior (model) on cpu/gpu
We should allow the ability to load the model on cpu/gpu

### LINKED ISSUE #615: Index mismatch when using `load_posterior`
When loading a dataset, the int labels of the saved dataset do not systematically match the original data. This is problematic concerning issue #606. 
This situation could, for instance, be avoided by systematically sorting the categorical variables.

### PR DESCRIPTION
**Pull request by PierreBoyeau**
fixes #615, fixes #612, fixes #606

### CODE REVIEW

File: scvi/inference/posterior.py [hunk 1]
```diff
@@ -150,7 +148,17 @@ def save_posterior(self, dir_path: str):
...
+        with open(os.path.join(dir_path, "posterior_type.txt"), "w") as post_file:
+            post_file.write(self.posterior_type)
         torch.save(self.model.state_dict(), os.path.join(dir_path, "model_params.pt"))
+
+        # Saves posterior indices and kwargs that can easily be retrieved
+        data_loader_kwargs = pd.Series(
+            {key: vals for key, vals in self.data_loader_kwargs.items()}
+        )
+        data_loader_kwargs.to_hdf(
+            os.path.join(dir_path, "data_loader_kwargs.csv"), key="data_loader"
```
Comment (adamgayoso) [inline scvi/inference/posterior.py]: file extension should be h5 here?

File: scvi/inference/posterior.py [hunk 2]
```diff
@@ -150,7 +148,17 @@ def save_posterior(self, dir_path: str):
         anndata_dataset = self.gene_dataset.to_anndata()
 
         anndata_dataset.write(os.path.join(dir_path, "anndata_dataset.h5ad"))
+        with open(os.path.join(dir_path, "posterior_type.txt"), "w") as post_file:
+            post_file.write(self.posterior_type)
         torch.save(self.model.state_dict(), os.path.join(dir_path, "model_params.pt"))
+
+        # Saves posterior indices and kwargs that can easily be retrieved
+        data_loader_kwargs = pd.Series(
+            {key: vals for key, vals in self.data_loader_kwargs.items()}
```
Comment (adamgayoso) [inline scvi/inference/posterior.py]: does this store the sampler? "sampler" is one of the keys in data_loader_kwargs no? I thought we were trying to avoid storing the sampler.

File: scvi/inference/posterior.py [hunk 3]
```diff
@@ -871,7 +889,13 @@ def differential_expression_score(
         :param all_stats: whether additional metrics should be provided
         :\**kwargs: Other keywords arguments for `get_sample_scale()`
 
-        :return: Differential expression properties
+        :return: Differential expression properties. The most important keys are:
```
Comment (adamgayoso) [inline scvi/inference/posterior.py]: keys -> columns?

File: scvi/inference/posterior_utils.py [hunk 4]
```diff
@@ -1,13 +1,102 @@
...
+from scipy.stats import entropy, kde
+from sklearn.neighbors import KNeighborsRegressor, NearestNeighbors
+
+from scvi.dataset import AnnDatasetFromAnnData
+
+
+def load_posterior(
+    dir_path: str,
+    model: nn.Module,
+    use_cuda: Optional[Union[bool, str]] = "auto",
```
Comment (adamgayoso) [inline scvi/inference/posterior_utils.py]: what does "auto" do? It's not being tested. We should also specify what it does in the docstring.

File: scvi/inference/posterior_utils.py [hunk 5]
```diff
@@ -1,13 +1,102 @@
...
+    indices = np.load(file=indices_path)
+    data_loader_kwargs = pd.read_hdf(
+        data_loader_kwargs_path, key="data_loader"
+    ).to_dict()
+    my_post = post_class(
+        model=model,
+        gene_dataset=dataset,
+        shuffle=False,
+        indices=indices,
+        use_cuda="auto",
```
Comment (adamgayoso) [inline scvi/inference/posterior_utils.py]: This should be true or false based on the "auto" option

File: scvi/inference/posterior_utils.py [hunk 6]
```diff
@@ -1,13 +1,102 @@
...
+    from scvi.inference.total_inference import TotalPosterior
+    from scvi.inference.jvae_trainer import JPosterior
+    from scvi.inference.posterior import Posterior
+    from scvi.inference.annotation import AnnotationPosterior
+
+    post_type_path = os.path.join(dir_path, "posterior_type.txt")
+    dataset_path = os.path.join(dir_path, "anndata_dataset.h5ad")
+    model_path = os.path.join(dir_path, "model_params.pt")
+    indices_path = os.path.join(dir_path, "indices.npy")
+    data_loader_kwargs_path = os.path.join(dir_path, "data_loader_kwargs.csv")
```
Comment (adamgayoso) [inline scvi/inference/posterior_utils.py]: h5 file extension here (or whatever is conventional for hdf)

File: scvi/inference/posterior.py (hunk 2 above)
Comment (adamgayoso) [reply scvi/inference/posterior.py]: we should also avoid storing the "collate_fn" key.

File: scvi/inference/posterior.py [hunk 7]
```diff
@@ -871,7 +889,13 @@ def differential_expression_score(
...
+
+            - proba_de (probability of being differentially expressed in change mode)
```
Comment (adamgayoso) [inline scvi/inference/posterior.py]: let's put wrap the names with ``name`` to make it look nice in RTD. It's two of ` on each side.

File: scvi/inference/posterior_utils.py [hunk 8]
```diff
@@ -1,13 +1,102 @@
...
+def load_posterior(
+    dir_path: str,
+    model: nn.Module,
+    use_cuda: Optional[Union[bool, str]] = "auto",
+    **posterior_kwargs
+):
+    """Function to use in order to retrieve a posterior that was saved using the `save_posterior` method
+
+    Because of pytorch model loading usage, this function needs a scVI model object initialized with exact same parameters
+    that during training.
```
Comment (adamgayoso) [inline scvi/inference/posterior_utils.py]: Can we say something about how it will be sequential after loading?

File: tests/test_scvi.py [hunk 9]
```diff
@@ -138,6 +139,23 @@ def test_synthetic_1():
...
+    with tempfile.TemporaryDirectory() as temp_dir:
+        posterior_save_path = os.path.join(temp_dir, "posterior_data")
+        original_post = trainer_synthetic_svaec.labelled_set.sequential()
+        original_post.save_posterior(posterior_save_path)
+        new_svaec = SCANVI(
+            synthetic_dataset.nb_genes,
+            synthetic_dataset.n_batches,
+            synthetic_dataset.n_labels,
+        )
+        new_post = load_posterior(posterior_save_path, model=new_svaec, use_cuda=False)
```
Comment (adamgayoso) [inline tests/test_scvi.py]: we should test without use_cuda set
